components/esp_websocket_client/examples/pytest_websocket.py

```diff
@@ -0,0 +1,128 @@
+import os
+import random
+import re
+import socket
+import string
+from threading import Event, Thread
+import pytest
+
+from SimpleWebSocketServer import SimpleWebSocketServer, WebSocket
+from pytest_embedded import Dut
+
+
+def get_my_ip():
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    try:
+        # doesn't even have to be reachable
+        s.connect(('8.8.8.8', 1))
+        IP = s.getsockname()[0]
+    except Exception:
+        IP = '127.0.0.1'
+    finally:
+        s.close()
+    return IP
+
+
+class WebsocketTestEcho(WebSocket):
+
+    def handleMessage(self):
+        self.sendMessage(self.data)
+        print('Server sent: {}'.format(self.data))
+
+    def handleConnected(self):
+        print('Connection from: {}'.format(self.address))
+
+    def handleClose(self):
+        print('{} closed the connection'.format(self.address))
+
+
+# Simple Websocket server for testing purposes
+class Websocket(object):
+
+    def send_data(self, data):
+        for nr, conn in self.server.connections.items():
+            conn.sendMessage(data)
+
+    def run(self):
+        self.server = SimpleWebSocketServer('', self.port, WebsocketTestEcho)
+        while not self.exit_event.is_set():
+            self.server.serveonce()
+
+    def __init__(self, port):
+        self.port = port
+        self.exit_event = Event()
+        self.thread = Thread(target=self.run)
+        self.thread.start()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.exit_event.set()
+        self.thread.join(10)
+        if self.thread.is_alive():
+            Utility.console_log('Thread cannot be joined', 'orange')
+
+
+def test_examples_protocol_websocket(dut):
+    """
+    steps:
+      1. obtain IP address
+      2. connect to uri specified in the config
+      3. send and receive data
+    """
+    def test_echo(dut):
+        dut.expect('WEBSOCKET_EVENT_CONNECTED')
+        for i in range(0, 5):
+            dut.expect(re.compile(b'Received=hello (\\d)'))
+        print('All echos received')
+
+    def test_close(dut):
+        code = dut.expect(re.compile(b'WEBSOCKET: Received closed message with code=(\\d*)'))[0]
+        print('Received close frame with code {}'.format(code))
+
+    def test_recv_long_msg(dut, websocket, msg_len, repeats):
+        send_msg = ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase + string.digits) for _ in range(msg_len))
+    
+        for _ in range(repeats):
+            websocket.send_data(send_msg)
+    
+            recv_msg = ''
+            while len(recv_msg) < msg_len:
+                # Filter out color encoding
+                match = dut.expect(re.compile(b'Received=([a-zA-Z0-9]*).*'))[0].decode()[9:-5] # We exclude initial `Received=` and last 5 characters which where printed in esp client
```

Comment:
I'm sorry I still don't understand why we exclude the last 5 characters...?
*` `Received=`  -- Understand that, it's 9 characters from the start (but wouldn't it be easier to simply check for group[1] -- when you defined it as `([a-zA-Z0-9]*)` ?)
* But why 5 characters? Is there anything else the client types? (My only guess would be the terminal's colors and newline chars?)

Reply:
@david-cermak good point, easier to check group(1)  and yes the last 5 characters are `\x1b[0m\r` terminal's colors and newline characters.